### PR TITLE
Fix an issue where the All Domains flow wasn't taking into account sites with existing plan

### LINKED ISSUE #22268: All Domains: Existing site flow doesn't handle the case when the site already has a plan
Existing site flow doesn't handle the case when the site already has a plan

### Expected behavior

Direct domain purchase should be opened instead of purchase with a plan when the user has a plan

### Steps to reproduce the behavior

1. Log into the site that already has a plan
2. All Domains
3. "+"
4. Select a domain
5. Select "Existing WP.com" site
6. Flow is stuck in Plans page

##### Tested on [device], iOS [version], Jetpack iOS / WordPress iOS [version]

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/2faf0a55-4266-40cb-b226-0a7bd17f027b

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #22268

## Description

This PR resolves an issue where users purchasing a domain, linked to a site that already has a plan, through the All Domains flow get stuck  in the **Plans** flow.

## Test Instructions

**N.B: Sandboxing is required to test domain checkout**

#### Case 1: Site with Existing Plan

1. Log into the site that already has a plan
2. Navigate to Me > All Domains
3. Tap "+"
4. Select a domain
5. Tap "Choose Site" button
6. **Expect** the **Checkout** web view to appear
7. Purchase the domain to verify the checkout flow works.

#### Case 2: Site without plan ( Regression )

1. Log into the site that doesn't have a plan
2. Navigate to Me > All Domains
3. Tap "+"
4. Select a domain
5. Tap "Get Domain" button
6. **Expect** the **Plans** web view to appear
7. Select a plan, go verify the checkout flow works as expected.

#### Case 3: Domain with No-Site ( Regression )

1. Log into the site that already has a plan
2. Navigate to Me > All Domains
3. Tap "+"
4. Select a domain
5. Tap "Get Domain" button
6. **Expect** the **Checkout** web view to appear
7. Purchase the domain to verify the checkout flow works.

## Next
- Address the  `// TODO:` I added in `RegisterDomainCoordinator.swift`.
- Add Unit Tests for `RegisterDomainCoordinator`.

## Regression Notes
1. Potential unintended areas of impact
Purchasing a domain through the All Domains flow.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 1]
```diff
@@ -141,20 +134,39 @@ class RegisterDomainCoordinator {
 
     // MARK: Helpers
 
+    private func purchaseDomain(for site: Blog?, on viewController: UIViewController, completion: @escaping (Result<Void, Swift.Error>) -> Void) {
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:151]: I think a VM would be a better fit for these service calls but it doesn't have to be part of this PR.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 2]
```diff
@@ -141,20 +134,39 @@ class RegisterDomainCoordinator {
...
+            switch result {
+            case .success(let domain):
+                if let site, site.canRegisterDomainWithPaidPlan {
+                    self.domainAddedToCartAndLinkedToSiteCallback?(viewController, domain.domainName, site)
+                } else {
+                    self.presentCheckoutWebview(on: viewController, title: TextContent.checkoutTitle)
+                }
+                completion(.success(()))
+            case .failure(let error):
+                viewController.displayActionableNotice(title: TextContent.errorTitle, actionTitle: TextContent.errorDismiss)
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:174]: Shouldn't the `completion` block be called on an action of `displayActionableNotice`?

**Comment by hassaanelgarem** [comment]
@salimbraksa Can you briefly explain how these changes fixed the issue? 🙏

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 3]
```diff
@@ -36,11 +41,13 @@ class RegisterDomainCoordinator {
...
          domainPurchasedCallback: RegisterDomainCoordinator.DomainPurchasedCallback? = nil,
          analyticsSource: String = "domains_register",
-         crashLogger: CrashLogging = .main) {
+         crashLogger: CrashLogging = .main,
+         domainRegistrationService: RegisterDomainDetailsServiceProxyProtocol = RegisterDomainDetailsServiceProxy()) {
         self.site = site
         self.domainPurchasedCallback = domainPurchasedCallback
         self.crashLogger = crashLogger
         self.analyticsSource = analyticsSource
+        self.domainRegistrationService = domainRegistrationService
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:52]: This service was instantiated within the `createCart` method, but I've moved to `init` for better dependency injection, allowing easier unit testing.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 4]
```diff
@@ -50,6 +57,7 @@ class RegisterDomainCoordinator {
     func handlePurchaseDomainOnly(on viewController: UIViewController,
                                   onSuccess: @escaping () -> (),
                                   onFailure: @escaping () -> ()) {
+        // TODO: Refactor `handlePurchaseDomainOnly` to use `purchaseDomain` helper.
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:62]: Will be addressed in another PR.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 5]
```diff
@@ -66,6 +74,7 @@ class RegisterDomainCoordinator {
     func addDomainToCartLinkedToCurrentSite(on viewController: UIViewController,
                          onSuccess: @escaping () -> (),
                          onFailure: @escaping () -> ()) {
+        // TODO: Refactor `addDomainToCartLinkedToCurrentSite` to use `purchaseDomain` helper.
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:79]: Will be addressed in another PR.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift (hunk 2 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:174]: Resolved in a DM.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 6]
```diff
@@ -141,20 +134,39 @@ class RegisterDomainCoordinator {
...
+            guard let self else {
+                return
+            }
+            switch result {
+            case .success(let domain):
+                if let site, site.canRegisterDomainWithPaidPlan {
+                    self.domainAddedToCartAndLinkedToSiteCallback?(viewController, domain.domainName, site)
+                } else {
+                    self.presentCheckoutWebview(on: viewController, title: TextContent.checkoutTitle)
+                }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: This is exactly the change that fixes the issue:
1. If `site.canRegisterDomainWithPaidPlan` is `true`, then we show the Plans web view.
2. Else, the Checkout web view is shown.

cc @hassaanelgarem

Comment (hassaanelgarem) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: Thanks for clarifying 🙏

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: @staskus Do you confirm [this](https://github.com/wordpress-mobile/WordPress-iOS/pull/22343#discussion_r1455397204) logic is correct? 🙏

Comment (staskus) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: I haven't checked the rest of the PR but if `site.canRegisterDomainWithPaidPlan` it means that the site has a domain credit, therefore we just allow to make domain registration without any plan selection or checkout:

image

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: @staskus Re-reading  `SiteDomainsViewModel` and `SiteDomainsView` logic, I understand that:
- The `Register domain` screen is shown when `site.canRegisterDomainWithPaidPlan` is `true`.
- The Plans web view is shown when `site.canRegisterDomainWithPaidPlan` is `false` and user doesn't have domains.
- The checkout web view is shown when `site.canRegisterDomainWithPaidPlan` is `false` and user has domains.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift [hunk 7]
```diff
@@ -13,16 +13,23 @@ class RegisterDomainCoordinator {
...
+    private let domainRegistrationService: RegisterDomainDetailsServiceProxyProtocol
+
     // MARK: Variables
 
     private let crashLogger: CrashLogging
 
     let analyticsSource: String
 
+    // TODO: This can cause Core Data crashes. Pass `NSManagedObjectID` instead.
     var site: Blog?
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:26]: FYI @hassaanelgarem I'll address this in a separate PR.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift (hunk 6 above)
Comment (staskus) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift:171]: Yes, sounds right!

**Comment by salimbraksa** [comment]
Following Povilas' [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/22343#discussion_r1455544434), it appears these changes don't resolve the issue. I'm converting this PR to a Draft and deprioritizing it to focus on the [In-App Feedback Prompt UI ](https://github.com/wordpress-mobile/WordPress-iOS/pull/22052)instead.
